src/Microsoft.AspNet.OData.Shared/TypeHelper.cs

```diff
@@ -491,5 +491,75 @@ private static Type GetInnerGenericType(Type interfaceType)
 
             return null;
         }
+
+        /// <summary>
+        /// Searches the public property with the specified name from the target type and its base types safely,
+        /// circumventing <see cref="AmbiguousMatchException"/> where a property has been redeclared in a derived type.
+        /// </summary>
+        /// <param name="targetType">The type to search from.</param>
+        /// <param name="propertyName">The name of the public property to get.</param>
+        /// <returns>
+        /// An object representing the public property with the specified name from the most derived class,
+        /// if found; otherwise, null.
+        /// </returns>
+        internal static PropertyInfo GetProperty(Type targetType, string propertyName)
+        {
+            // NOTE: Designed to avoid costly closure lambda allocations since method might be used in hot path
+            List<PropertyInfo> matchedProperties = null;
+
+            PropertyInfo[] properties = targetType.GetProperties();
+            for (int i = 0; i < properties.Length; i++)
+            {
+                PropertyInfo propertyInfo = properties[i];
```

Comment:
Since `properties` is an array, I think using `foreach` is better since it will be compiled down to the same thing.

Reply:
@habbes Since the for loop in this case is fairly straightforward, also owing to the fact that not everyone will be familiar with the compiler optimization, I'm inclined towards leave it be. I have however implemented the optimization fix such that we don't initialize the list unless 2 or more properties are matched